stdlib/Random/src/generation.jl

```diff
@@ -167,6 +167,11 @@ function rand(r::AbstractRNG, ::SamplerType{T}) where {T<:AbstractChar}
     (c < 0xd800) ? T(c) : T(c+0x800)
 end
 
+### random tuples
+
+function rand(r::AbstractRNG, ::SamplerType{T}) where {T<:Tuple}
+    ntuple(i -> rand(r, T.parameters[i]), length(T.parameters))
+end
```

Comment:
Is this guaranteed to be type stable?

I wonder if we should use a generated function instead.

Reply:
good point. I see `ntuple` allocates past `10` parameters, I can try the `Val{N}` version. also I didn't check that this works with `Vararg` so there might be a little more work to do there

Reply:
I think this is basically what the `Val` version does, under an `if @generated` branch

```
    if @generated
        :(@ntuple $N i -> f(i))
```

where
```
macro ntuple(N::Int, ex)
    vars = Any[ inlineanonymous(ex,i) for i = 1:N ]
    Expr(:escape, Expr(:tuple, vars...))
end
```

I found in the `Metaprogramming` section the guidance `This [optionally @generated] style is preferred, since it allows the compiler to make more decisions and compile
programs in more ways`, hence my option to use the existing `ntuple(, ::Val)`

Anyway, I tried both versions on a wide tuple (~200 parameters) and they had nearly identical performance within a few nanos :)

Reply:
This should not access `.parameters` (which a private, possibly unstable, API), but rather map over the `fieldtype(T, i)`

Reply:
@vtjnash how about `ntuple(i -> rand(r, fieldtype(T, i)), Val(fieldcount(T)))` ?

Reply:
`map(t -> rand(r, t), fieldtypes(T))`?

Reply:
or just `rand.(r, fieldtypes(T))`?

Reply:
thank you both for the suggestions. both of those alternate implementations are indeed easier to read, but unfortunately they seem to have worse performance on wide & nested types

```
### broadcasting
julia> @btime rand(Tuple{Bool, Char, Vararg{Tuple{Int, Float64, Tuple{Char, Bool}}, 80}})
  46.417 μs (486 allocations: 12.77 KiB)


### map
julia> @btime rand(Tuple{Bool, Char, Vararg{Tuple{Int, Float64, Tuple{Char, Bool}}, 80}})
  10.208 μs (85 allocations: 6.59 KiB)


### this PR
julia> @btime rand(Tuple{Bool, Char, Vararg{Tuple{Int, Float64, Tuple{Char, Bool}}, 80}})
  1.212 μs (80 allocations: 2.50 KiB)
```